Defining libraries

In [2]:
import pandas as pd
import glob
import numpy as np
import matplotlib as plt
import sqlalchemy as sa
import datetime

Process the csv's files

In [4]:
files = glob.glob("JC-2016*-citibike-tripdata.csv")

all_files = []
for csv in files:
    data = pd.read_csv(csv)
    all_files.append(data)
df = pd.concat(all_files)

In [5]:
df.head()

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            362  2016-01-01 00:02:52  2016-01-01 00:08:54              3186   
1            200  2016-01-01 00:18:22  2016-01-01 00:21:42              3186   
2            202  2016-01-01 00:18:25  2016-01-01 00:21:47              3186   
3            248  2016-01-01 00:23:13  2016-01-01 00:27:21              3209   
4            903  2016-01-01 01:03:20  2016-01-01 01:18:24              3195   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   End Station ID End Station Name  End Station Latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  
0             -74.050656    24647  Subscriber      1964.0       2  
1             -74.047727    24605  Subscriber      1962.0       1  
2             -74.047727    24689  Subscriber      1962.0       2  
3             -74.044247    24693  Subscriber      1984.0       1  
4             -74.051789    24573    Customer         NaN       0

In [6]:
df['id'] = df.index
df.head()

Trip Duration           Start Time            Stop Time  Start Station ID  \
0            362  2016-01-01 00:02:52  2016-01-01 00:08:54              3186   
1            200  2016-01-01 00:18:22  2016-01-01 00:21:42              3186   
2            202  2016-01-01 00:18:25  2016-01-01 00:21:47              3186   
3            248  2016-01-01 00:23:13  2016-01-01 00:27:21              3209   
4            903  2016-01-01 01:03:20  2016-01-01 01:18:24              3195   

  Start Station Name  Start Station Latitude  Start Station Longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   End Station ID End Station Name  End Station Latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   End Station Longitude  Bike ID   User Type  Birth Year  Gender  id  
0             -74.050656    24647  Subscriber      1964.0       2   0  
1             -74.047727    24605  Subscriber      1962.0       1   1  
2             -74.047727    24689  Subscriber      1962.0       2   2  
3             -74.044247    24693  Subscriber      1984.0       1   3  
4             -74.051789    24573    Customer         NaN       0   4

- Observation
    1. User Type only has 2 kind of values. Subscriber and Customer
    2. We can see that Birth Year has missing data
    3. Gender has 3 kind of values : Zero = unknown, 1=male, 2=female 

See the information of the dataset

In [9]:
#wathc the infor of our dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 247584 entries, 0 to 15113
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Trip Duration            247584 non-null  int64  
 1   Start Time               247584 non-null  object 
 2   Stop Time                247584 non-null  object 
 3   Start Station ID         247584 non-null  int64  
 4   Start Station Name       247584 non-null  object 
 5   Start Station Latitude   247584 non-null  float64
 6   Start Station Longitude  247584 non-null  float64
 7   End Station ID           247584 non-null  int64  
 8   End Station Name         247584 non-null  object 
 9   End Station Latitude     247584 non-null  float64
 10  End Station Longitude    247584 non-null  float64
 11  Bike ID                  247584 non-null  int64  
 12  User Type                247204 non-null  object 
 13  Birth Year               228585 non-null  float64
 14  Gender    

- Observations
  1. Start time and Stop Time have an object type insted datatime type
  2. Birth Year have a float64 type insted int64 type
  3. We have 247584 rows in our dataframe 


In [11]:
#watch the description of our dataset
df.describe()

Trip Duration  Start Station ID  Start Station Latitude  \
count   2.475840e+05     247584.000000           247584.000000   
mean    8.856305e+02       3207.065206               40.723121   
std     3.593798e+04         26.955103                0.008199   
min     6.100000e+01       3183.000000               40.692640   
25%     2.480000e+02       3186.000000               40.717732   
50%     3.900000e+02       3201.000000               40.721525   
75%     6.660000e+02       3211.000000               40.727596   
max     1.632981e+07       3426.000000               40.752559   

       Start Station Longitude  End Station ID  End Station Latitude  \
count            247584.000000   247584.000000         247584.000000   
mean                -74.046438     3203.572553             40.722594   
std                   0.011211       61.579494              0.007958   
min                 -74.096937      147.000000             40.692216   
25%                 -74.050656     3186.000000             40.716540   
50%                 -74.044247     3199.000000             40.721124   
75%                 -74.038051     3211.000000             40.727224   
max                 -74.032108     3426.000000             40.801343   

       End Station Longitude        Bike ID     Birth Year         Gender  \
count          247584.000000  247584.000000  228585.000000  247584.000000   
mean              -74.045855   24935.260481    1979.335276       1.123534   
std                 0.011283     748.469712       9.596809       0.518687   
min               -74.096937   14552.000000    1900.000000       0.000000   
25%               -74.050444   24491.000000    1974.000000       1.000000   
50%               -74.043117   24609.000000    1981.000000       1.000000   
75%               -74.036486   24719.000000    1986.000000       1.000000   
max               -73.957390   27274.000000    2000.000000       2.000000   

                  id  
count  247584.000000  
mean    12094.850350  
std      8356.014112  
min         0.000000  
25%      5157.750000  
50%     10806.000000  
75%     17855.250000  
max     34148.000000

- Observation
    1. Trip Duration has an rare increment in 75% of 3 units.
    2. Biker ID the min value is 1900. 116 years?

# Starting with Start time and End Time

In [14]:
df.columns

Index(['Trip Duration', 'Start Time', 'Stop Time', 'Start Station ID',
       'Start Station Name', 'Start Station Latitude',
       'Start Station Longitude', 'End Station ID', 'End Station Name',
       'End Station Latitude', 'End Station Longitude', 'Bike ID', 'User Type',
       'Birth Year', 'Gender', 'id'],
      dtype='object')

In [15]:
#Change the name of the columns
df.rename(columns = {'Trip Duration': 'trip_duration', 
                     'Start Time': 'start_time', 
                     'Stop Time': 'stop_time', 
                     'Start Station ID':'start_station_id',
                     'Start Station Name': 'start_station_name',
                     'Start Station Latitude':'start_station_latitude',
                     'Start Station Longitude':'start_station_longitude',
                     'End Station ID':'end_station_id',
                     'End Station Name':'end_station_name',
                     'End Station Latitude':'end_station_latitude',
                     'End Station Longitude':'end_station_longitude',
                     'Bike ID':'bike_id',
                     'User Type':'user_type',
                     'Birth Year':'birth_year',
                     'Gender':'gender'
                    }, inplace = True)

In [16]:
df['start_time'] = pd.to_datetime(df['start_time'])
df['stop_time'] = pd.to_datetime(df['stop_time'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 247584 entries, 0 to 15113
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   trip_duration            247584 non-null  int64         
 1   start_time               247584 non-null  datetime64[ns]
 2   stop_time                247584 non-null  datetime64[ns]
 3   start_station_id         247584 non-null  int64         
 4   start_station_name       247584 non-null  object        
 5   start_station_latitude   247584 non-null  float64       
 6   start_station_longitude  247584 non-null  float64       
 7   end_station_id           247584 non-null  int64         
 8   end_station_name         247584 non-null  object        
 9   end_station_latitude     247584 non-null  float64       
 10  end_station_longitude    247584 non-null  float64       
 11  bike_id                  247584 non-null  int64         
 12  user_type             

# Missing Data information

In [18]:
#count the missing data from of all the columns
missing_data_df = df.isnull().sum()
missing_data_df

trip_duration                  0
start_time                     0
stop_time                      0
start_station_id               0
start_station_name             0
start_station_latitude         0
start_station_longitude        0
end_station_id                 0
end_station_name               0
end_station_latitude           0
end_station_longitude          0
bike_id                        0
user_type                    380
birth_year                 18999
gender                         0
id                             0
dtype: int64

- Observation
    1. User Type has 380 missing data values
    2. Birth Year has 18999 missing data values

Analyze the Birth Year min value

In [21]:
df['age'] = 2016 - df['birth_year']
df['age'].sort_values(ascending=False)[:10]

4417     116.0
22758     82.0
16672     79.0
6163      79.0
32771     79.0
2929      79.0
29718     76.0
23319     76.0
2201      76.0
11048     75.0
Name: age, dtype: float64

We can see that the low value have the index 4417 that is the "person" that have 116 year. In this case is recomended delete that "person" 

In [23]:
df.drop([4417], inplace = True)
df['age'].sort_values(ascending=False)[:10]

22758    82.0
6163     79.0
2929     79.0
16672    79.0
32771    79.0
23319    76.0
2201     76.0
29718    76.0
33877    75.0
30379    75.0
Name: age, dtype: float64

Investigate the missing values from birth year

In [25]:
missing_data_by = df[df[['birth_year']].isnull().any(axis = 1)]
missing_data_by

trip_duration          start_time           stop_time  \
4                903 2016-01-01 01:03:20 2016-01-01 01:18:24   
5                883 2016-01-01 01:03:28 2016-01-01 01:18:11   
22               988 2016-01-01 03:16:33 2016-01-01 03:33:02   
53              3090 2016-01-01 11:07:15 2016-01-01 11:58:46   
57               788 2016-01-01 11:50:30 2016-01-01 12:03:39   
...              ...                 ...                 ...   
14972           2484 2016-12-31 14:35:19 2016-12-31 15:16:44   
14973          79669 2016-12-31 14:35:25 2017-01-01 12:43:14   
15067            224 2016-12-31 18:01:38 2016-12-31 18:05:23   
15091            440 2016-12-31 20:12:26 2016-12-31 20:19:47   
15110           2749 2016-12-31 23:29:39 2017-01-01 00:15:29   

       start_station_id start_station_name  start_station_latitude  \
4                  3195            Sip Ave               40.730743   
5                  3195            Sip Ave               40.730743   
22                 3196     Riverview Park               40.744319   
53                 3203      Hamilton Park               40.727596   
57                 3210     Pershing Field               40.742677   
...                 ...                ...                     ...   
14972              3275     Columbus Drive               40.718355   
14973              3275     Columbus Drive               40.718355   
15067              3194    McGinley Square               40.725340   
15091              3195            Sip Ave               40.730743   
15110              3183     Exchange Place               40.716247   

       start_station_longitude  end_station_id end_station_name  \
4                   -74.063784            3210   Pershing Field   
5                   -74.063784            3210   Pershing Field   
22                  -74.043991            3209     Brunswick St   
53                  -74.044247            3203    Hamilton Park   
57                  -74.051789            3195          Sip Ave   
...                        ...             ...              ...   
14972               -74.038914            3199     Newport Pkwy   
14973               -74.038914            3199     Newport Pkwy   
15067               -74.067622            3195          Sip Ave   
15091               -74.063784            3193     Lincoln Park   
15110               -74.033459            3183   Exchange Place   

       end_station_latitude  end_station_longitude  bike_id   user_type  \
4                 40.742677             -74.051789    24573    Customer   
5                 40.742677             -74.051789    24442    Customer   
22                40.724176             -74.050656    24662    Customer   
53                40.727596             -74.044247    24444    Customer   
57                40.730743             -74.063784    24573    Customer   
...                     ...                    ...      ...         ...   
14972             40.728745             -74.032108    24627    Customer   
14973             40.728745             -74.032108    26217    Customer   
15067             40.730743             -74.063784    24716  Subscriber   
15091             40.724605             -74.078406    24417  Subscriber   
15110             40.716247             -74.033459    24389    Customer   

       birth_year  gender     id  age  
4             NaN       0      4  NaN  
5             NaN       0      5  NaN  
22            NaN       0     22  NaN  
53            NaN       0     53  NaN  
57            NaN       0     57  NaN  
...           ...     ...    ...  ...  
14972         NaN       0  14972  NaN  
14973         NaN       0  14973  NaN  
15067         NaN       0  15067  NaN  
15091         NaN       0  15091  NaN  
15110         NaN       0  15110  NaN  

[18999 rows x 17 columns]

- Observation
    1. There could be a related between user_type, birth_year and gender, this becouse the nan values from birth_year match with 0 values 
from gender and some customers from user_type
    2. Customer from user_type will be related to Nan data from birth_year?

In [27]:
#Count the customer data from user_type that are releted with Nan data from birth_year
count_missing_data_ut = missing_data_by['user_type'].value_counts(dropna = False)
count_missing_data_ut

user_type
Customer      15470
Subscriber     3529
Name: count, dtype: int64

- Observation
    1. We can see that it has 15470 from customer that have no value for the birth year
    2. We can see that is has 3529 from Subscriber that have not value for the birth year

In [29]:
#Count all the values from the original dataframe
count_ut_values = df['user_type'].value_counts(dropna = False) 
count_ut_values

user_type
Subscriber    231671
Customer       15521
NaN              380
Name: count, dtype: int64

In [30]:
# see how many Customers and Subscribers are related with the missing data from Birth Year
((count_ut_values - count_missing_data_ut) / count_ut_values) * 100

user_type
Customer       0.328587
Subscriber    98.476719
NaN                 NaN
Name: count, dtype: float64

- Observation
    1. We can see that only 51 customer that represent the 0.32 % have information within Birth Year
    2. In subscriber we have 228142 that represent the 98.46% of data have information within Birth Year

That tell us it's not possible drop the missing data form Customers because only the 0.32 % of the customers have information in Birth Day
In this case we need write a report about that.
We will have to keep this missing data in mind when we developing the database.

Missing data from User Type

In [34]:
#see if the missing data from User Type have a related with birth year
mdut = df[df['user_type'].isna()]
(mdut['birth_year'].isna()).sum()

0

We can see that missing data from User Type have no relation with missing data from Birth Year

Now we will to see if a missing data from gender(0) are releted with missing data from User Type 

In [37]:
(mdut['gender'] == 0).sum() # gender has not Nan values, it's has 0 values

0

We can see neither of the two are related with missing data from user_type.
In this case maybe the missing data from User Type wont be Missing at random.
For good practice wi will change the 0 from Gender to unknown.

In [39]:
df.fillna({'user_type':'Unknown'},inplace=True)
df['user_type'].value_counts()

user_type
Subscriber    231671
Customer       15521
Unknown          380
Name: count, dtype: int64

In [40]:
df.head(10)

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   
5            883 2016-01-01 01:03:28 2016-01-01 01:18:11              3195   
6            445 2016-01-01 01:07:45 2016-01-01 01:15:11              3186   
7            192 2016-01-01 01:18:51 2016-01-01 01:22:03              3211   
8            409 2016-01-01 01:23:44 2016-01-01 01:30:34              3187   
9            285 2016-01-01 01:25:12 2016-01-01 01:29:57              3187   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   
5            Sip Ave               40.730743               -74.063784   
6      Grove St PATH               40.719586               -74.043117   
7         Newark Ave               40.721525               -74.046305   
8          Warren St               40.721124               -74.038051   
9          Warren St               40.721124               -74.038051   

   end_station_id  end_station_name  end_station_latitude  \
0            3209      Brunswick St             40.724176   
1            3213    Van Vorst Park             40.718489   
2            3213    Van Vorst Park             40.718489   
3            3203     Hamilton Park             40.727596   
4            3210    Pershing Field             40.742677   
5            3210    Pershing Field             40.742677   
6            3203     Hamilton Park             40.727596   
7            3203     Hamilton Park             40.727596   
8            3214  Essex Light Rail             40.712774   
9            3214  Essex Light Rail             40.712774   

   end_station_longitude  bike_id   user_type  birth_year  gender  id   age  
0             -74.050656    24647  Subscriber      1964.0       2   0  52.0  
1             -74.047727    24605  Subscriber      1962.0       1   1  54.0  
2             -74.047727    24689  Subscriber      1962.0       2   2  54.0  
3             -74.044247    24693  Subscriber      1984.0       1   3  32.0  
4             -74.051789    24573    Customer         NaN       0   4   NaN  
5             -74.051789    24442    Customer         NaN       0   5   NaN  
6             -74.044247    24510  Subscriber      1988.0       2   6  28.0  
7             -74.044247    24625  Subscriber      1980.0       1   7  36.0  
8             -74.036486    24429  Subscriber      1990.0       1   8  26.0  
9             -74.036486    24407  Subscriber      1988.0       2   9  28.0

Gender

In [42]:
#see if gender has missing data 
df['gender'].isna().sum()

0

In [43]:
#See how many customers and suscriber there are in differents genders
df.groupby(['user_type','gender']).count()['bike_id']

user_type   gender
Customer    0          15470
            1             36
            2             15
Subscriber  0           4431
            1         176890
            2          50350
Unknown     1            262
            2            118
Name: bike_id, dtype: int64

- Observation
    1. The customers has 15470 data with 0 value in gender
    2. The subscriber has 4431 data with 0 value in gender
We can see that customers have 99% of data with 0 values in gender
We wont to remove customers, but we only need to now that the useful information is in subscriber 

In [45]:
df['id'].reset_index(drop=True)

0             0
1             1
2             2
3             3
4             4
          ...  
247567    15109
247568    15110
247569    15111
247570    15112
247571    15113
Name: id, Length: 247572, dtype: int64

In [46]:
df.head()

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end_station_id end_station_name  end_station_latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   end_station_longitude  bike_id   user_type  birth_year  gender  id   age  
0             -74.050656    24647  Subscriber      1964.0       2   0  52.0  
1             -74.047727    24605  Subscriber      1962.0       1   1  54.0  
2             -74.047727    24689  Subscriber      1962.0       2   2  54.0  
3             -74.044247    24693  Subscriber      1984.0       1   3  32.0  
4             -74.051789    24573    Customer         NaN       0   4   NaN

* Trip Duration
    - Trip Duration has an rare increment in 75% of 3 units.
    - The trip duration are in seconds

We will add some columns to have more information about the time

In [49]:
#add two columns more: trip_duration_minutes and trip_duration_hours
df['trip_duration_minutes'] = round(df['trip_duration'].apply(lambda x : x / 60), 2)
df['trip_duration_hours'] = round(df['trip_duration'].apply(lambda x : x / 60**2), 2)
df.head()

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end_station_id end_station_name  end_station_latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   end_station_longitude  bike_id   user_type  birth_year  gender  id   age  \
0             -74.050656    24647  Subscriber      1964.0       2   0  52.0   
1             -74.047727    24605  Subscriber      1962.0       1   1  54.0   
2             -74.047727    24689  Subscriber      1962.0       2   2  54.0   
3             -74.044247    24693  Subscriber      1984.0       1   3  32.0   
4             -74.051789    24573    Customer         NaN       0   4   NaN   

   trip_duration_minutes  trip_duration_hours  
0                   6.03                 0.10  
1                   3.33                 0.06  
2                   3.37                 0.06  
3                   4.13                 0.07  
4                  15.05                 0.25

In [50]:
df.describe()

trip_duration                     start_time  \
count   2.475720e+05                         247572   
mean    8.856405e+02  2016-07-29 05:57:31.958601984   
min     6.100000e+01            2016-01-01 00:02:52   
25%     2.480000e+02     2016-05-27 07:46:11.500000   
50%     3.900000e+02            2016-08-10 09:25:31   
75%     6.660000e+02  2016-10-05 17:25:37.750000128   
max     1.632981e+07            2016-12-31 23:44:50   
std     3.593885e+04                            NaN   

                           stop_time  start_station_id  \
count                         247572     247572.000000   
mean   2016-07-29 06:12:18.098290432       3207.065318   
min              2016-01-01 00:08:54       3183.000000   
25%              2016-05-27 07:56:36       3186.000000   
50%       2016-08-10 09:36:25.500000       3201.000000   
75%              2016-10-05 17:34:12       3211.000000   
max              2017-01-18 14:26:46       3426.000000   
std                              NaN         26.955263   

       start_station_latitude  start_station_longitude  end_station_id  \
count           247572.000000            247572.000000   247572.000000   
mean                40.723121               -74.046438     3203.572516   
min                 40.692640               -74.096937      147.000000   
25%                 40.717732               -74.050656     3186.000000   
50%                 40.721525               -74.044247     3199.000000   
75%                 40.727596               -74.038051     3211.000000   
max                 40.752559               -74.032108     3426.000000   
std                  0.008199                 0.011211       61.580654   

       end_station_latitude  end_station_longitude        bike_id  \
count         247572.000000          247572.000000  247572.000000   
mean              40.722594             -74.045855   24935.264279   
min               40.692216             -74.096937   14552.000000   
25%               40.716540             -74.050444   24491.000000   
50%               40.721124             -74.043117   24609.000000   
75%               40.727224             -74.036486   24719.000000   
max               40.801343             -73.957390   27274.000000   
std                0.007958               0.011283     748.475924   

          birth_year         gender             id            age  \
count  228573.000000  247572.000000  247572.000000  228573.000000   
mean     1979.335451       1.123528   12095.222501      36.664549   
min      1934.000000       0.000000       0.000000      16.000000   
25%      1974.000000       1.000000    5158.000000      30.000000   
50%      1981.000000       1.000000   10806.000000      35.000000   
75%      1986.000000       1.000000   17856.000000      42.000000   
max      2000.000000       2.000000   34148.000000      82.000000   
std         9.595412       0.518690    8356.045641       9.595412   

       trip_duration_minutes  trip_duration_hours  
count          247572.000000        247572.000000  
mean               14.760669             0.246009  
min                 1.020000             0.020000  
25%                 4.130000             0.070000  
50%                 6.500000             0.110000  
75%                11.100000             0.180000  
max            272163.470000          4536.060000  
std               598.980795             9.983019

* 0bservations
  - We can see the minimun trip duration is 1 min and de maximun trip duration were 4536 hours?
Maybe exist more values that exceed 24 hours

In [52]:
df['trip_duration_hours'].apply(lambda x : 1 if x > 24 else 0).sum()

93

- We can see that exist 93 bikers that exceed the 24(86400 seconds) hours of trip, that is really weird
- We won't remove becouse maybe it could be an error, only we will put a flag in each biker that exceed the 24 hours 

In [54]:
df['trip_flag'] = df['trip_duration'].apply(lambda x: True if x < 8660 else False).astype(bool)
df.head()

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end_station_id end_station_name  end_station_latitude  \
0            3209     Brunswick St             40.724176   
1            3213   Van Vorst Park             40.718489   
2            3213   Van Vorst Park             40.718489   
3            3203    Hamilton Park             40.727596   
4            3210   Pershing Field             40.742677   

   end_station_longitude  bike_id   user_type  birth_year  gender  id   age  \
0             -74.050656    24647  Subscriber      1964.0       2   0  52.0   
1             -74.047727    24605  Subscriber      1962.0       1   1  54.0   
2             -74.047727    24689  Subscriber      1962.0       2   2  54.0   
3             -74.044247    24693  Subscriber      1984.0       1   3  32.0   
4             -74.051789    24573    Customer         NaN       0   4   NaN   

   trip_duration_minutes  trip_duration_hours  trip_flag  
0                   6.03                 0.10       True  
1                   3.33                 0.06       True  
2                   3.37                 0.06       True  
3                   4.13                 0.07       True  
4                  15.05                 0.25       True

# Weather Dataset

In [56]:
df_w = pd.read_csv("newark_airport_2016.csv")
df_w.head(10)

STATION                                         NAME        DATE  \
0  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-01   
1  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-02   
2  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-03   
3  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-04   
4  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-05   
5  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-06   
6  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-07   
7  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-08   
8  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-09   
9  USW00014734  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US  2016-01-10   

    AWND  PGTM  PRCP  SNOW  SNWD  TAVG  TMAX  TMIN  TSUN  WDF2   WDF5  WSF2  \
0  12.75   NaN  0.00   0.0   0.0    41    43    34   NaN   270  280.0  25.9   
1   9.40   NaN  0.00   0.0   0.0    36    42    30   NaN   260  260.0  21.0   
2  10.29   NaN  0.00   0.0   0.0    37    47    28   NaN   270  250.0  23.9   
3  17.22   NaN  0.00   0.0   0.0    32    35    14   NaN   330  330.0  25.9   
4   9.84   NaN  0.00   0.0   0.0    19    31    10   NaN   360  350.0  25.1   
5   5.37   NaN  0.00   0.0   0.0    28    42    15   NaN   230  250.0  12.1   
6   3.36   NaN  0.00   0.0   0.0    35    46    24   NaN    20  360.0   8.9   
7   8.05   NaN  0.00   0.0   0.0    38    45    31   NaN    20   30.0  14.1   
8   6.71   NaN  0.01   0.0   0.0    44    48    38   NaN    60   70.0  13.0   
9  15.43   NaN  1.77   0.0   0.0    53    65    39   NaN   260  270.0  36.0   

   WSF5  
0  35.1  
1  25.1  
2  30.0  
3  33.1  
4  31.1  
5  16.1  
6  10.1  
7  16.1  
8  17.0  
9  42.9

- Description
    1. AWND = Average daily wind speed
    2. PGTM = Peak gust time
    3. PRCP = Precipitation
    4. SNOW = Snowfall
    5. SNWD = Snow depth
    6. TAVG = Temperature AVG
    7. TMAX = Maximum temperature
    8. TMIN = Minimum temperature
    9. TSUN = Daily total sunshine
    10. WDF2 = Direction of fastest 2-minute wind
    11. WDF5 = Direction of fastest 5-second wind
    12. WSF2 Fastest 2-minute wind speed
    13. WSF5 = Fastest 5-second wind speed

- Observation
    1. Station and Name have the same information
    2. PGTM and TSUN have missing data, maybe we'll drop those columns

In [59]:
# See the information of the dataset
df_w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 16 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   STATION  366 non-null    object 
 1   NAME     366 non-null    object 
 2   DATE     366 non-null    object 
 3   AWND     366 non-null    float64
 4   PGTM     0 non-null      float64
 5   PRCP     366 non-null    float64
 6   SNOW     366 non-null    float64
 7   SNWD     366 non-null    float64
 8   TAVG     366 non-null    int64  
 9   TMAX     366 non-null    int64  
 10  TMIN     366 non-null    int64  
 11  TSUN     0 non-null      float64
 12  WDF2     366 non-null    int64  
 13  WDF5     364 non-null    float64
 14  WSF2     366 non-null    float64
 15  WSF5     364 non-null    float64
dtypes: float64(9), int64(4), object(3)
memory usage: 45.9+ KB


- Observation
    1. Date has an object type insted datatime type
    2. PGTM and TSUM have no data, we can delete those columns
    3. WSF5 and WDF5 have missing data(only 2 missing data record)
    4. TAVG, TMAX and TMIN have float64 type but it need to have int64 data type

In [61]:
#See the description of the Dataframe
df_w.describe()

AWND  PGTM        PRCP        SNOW        SNWD        TAVG  \
count  366.000000   0.0  366.000000  366.000000  366.000000  366.000000   
mean     9.429973   NaN    0.104945    0.098087    0.342623   57.196721   
std      3.748174   NaN    0.307496    1.276498    2.078510   17.466981   
min      2.460000   NaN    0.000000    0.000000    0.000000    8.000000   
25%      6.765000   NaN    0.000000    0.000000    0.000000   43.000000   
50%      8.720000   NaN    0.000000    0.000000    0.000000   56.000000   
75%     11.410000   NaN    0.030000    0.000000    0.000000   74.000000   
max     22.820000   NaN    2.790000   24.000000   20.100000   89.000000   

             TMAX        TMIN  TSUN        WDF2        WDF5        WSF2  \
count  366.000000  366.000000   0.0  366.000000  364.000000  366.000000   
mean    65.991803   48.459016   NaN  217.841530  228.269231   20.484426   
std     18.606301   17.135790   NaN  102.548282   97.415777    6.848390   
min     18.000000    0.000000   NaN   10.000000   10.000000    6.900000   
25%     51.250000   35.000000   NaN  150.000000  150.000000   15.000000   
50%     66.000000   47.000000   NaN  240.000000  260.000000   19.900000   
75%     83.000000   64.000000   NaN  300.000000  300.000000   23.900000   
max     99.000000   80.000000   NaN  360.000000  360.000000   48.100000   

             WSF5  
count  364.000000  
mean    26.801648  
std      8.882610  
min     10.100000  
25%     19.900000  
50%     25.100000  
75%     31.100000  
max     66.000000

- Observation
    * All the data description looks good
    * It's important to undestend that WDF2, WDF5, WSF2 and WSF5 make refer to faster of the wind, this is not useful for the measurements that we want to do, wi will drop those columns.
    * We will drop PGTM and TSUN to.

In [63]:
df_w.drop(columns = ['STATION','PGTM', 'TSUN', 'WDF2', 'WDF5', 'WSF2', 'WSF5'], axis = 1, inplace = True)

In [64]:
df_w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   NAME    366 non-null    object 
 1   DATE    366 non-null    object 
 2   AWND    366 non-null    float64
 3   PRCP    366 non-null    float64
 4   SNOW    366 non-null    float64
 5   SNWD    366 non-null    float64
 6   TAVG    366 non-null    int64  
 7   TMAX    366 non-null    int64  
 8   TMIN    366 non-null    int64  
dtypes: float64(4), int64(3), object(2)
memory usage: 25.9+ KB


* Rename the columns

In [66]:
df_w.columns = [x.lower() for x in df_w.columns]
df_w.columns

Index(['name', 'date', 'awnd', 'prcp', 'snow', 'snwd', 'tavg', 'tmax', 'tmin'], dtype='object')

# Missing Data Analisys

In [68]:
#Changing Data type
df_w['date'] = pd.to_datetime(df_w['date'], yearfirst = True)
df_w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 366 entries, 0 to 365
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   name    366 non-null    object        
 1   date    366 non-null    datetime64[ns]
 2   awnd    366 non-null    float64       
 3   prcp    366 non-null    float64       
 4   snow    366 non-null    float64       
 5   snwd    366 non-null    float64       
 6   tavg    366 non-null    int64         
 7   tmax    366 non-null    int64         
 8   tmin    366 non-null    int64         
dtypes: datetime64[ns](1), float64(4), int64(3), object(1)
memory usage: 25.9+ KB


# Create tables

In [70]:
df.columns

Index(['trip_duration', 'start_time', 'stop_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'birth_year', 'gender', 'id', 'age', 'trip_duration_minutes',
       'trip_duration_hours', 'trip_flag'],
      dtype='object')

In [71]:
df_w.columns

Index(['name', 'date', 'awnd', 'prcp', 'snow', 'snwd', 'tavg', 'tmax', 'tmin'], dtype='object')

To create a database in postgres we need to create relational tables with the information of our database
- First we need to find a relation between all the tables, that relation is the "DATA"
    - We need to create a table with the time information
- Second we need to split the tables having the next schemas:
    - A table with the time informacion 
    - A table with the trip information
    - A table with the start_station information
    - A table with the end_stations information
    - A table with biker informacion
    - A table with weather informacion
- Third we need to find and drop the duplicate information

### Creating a data table

* Observation
    - Our data table will have year, month, day

In [75]:
#we use datatime library
datetime_init = datetime.datetime(2016,1,1)
data_colum = [datetime_init + datetime.timedelta(x) for x in range(366)]
#print(data_colum)
df_date = pd.DataFrame(data_colum, dtype = 'datetime64[ns]')
df_date.columns = ['date']
df_date.head(10)

date
0 2016-01-01
1 2016-01-02
2 2016-01-03
3 2016-01-04
4 2016-01-05
5 2016-01-06
6 2016-01-07
7 2016-01-08
8 2016-01-09
9 2016-01-10

We will include the next columns: date_key year, month and day. date_key will be our primary key 

In [77]:
df_date['id'] = df_date.index
df_date['date_key'] = df_date['date'].apply(lambda x: int(x.strftime('%Y%m%d')))
df_date['year'] = df_date['date'].apply(lambda x: x.year)
df_date['month'] = df_date['date'].apply(lambda x: x.month)
df_date['day'] = df_date['date'].apply(lambda x: x.day)
df_date.head()

date  id  date_key  year  month  day
0 2016-01-01   0  20160101  2016      1    1
1 2016-01-02   1  20160102  2016      1    2
2 2016-01-03   2  20160103  2016      1    3
3 2016-01-04   3  20160104  2016      1    4
4 2016-01-05   4  20160105  2016      1    5

Now we will create a data_key column in bike table and weather table

In [79]:
df['date_key'] = df['start_time'].apply(lambda x: int(x.strftime('%Y%m%d')))
df_w['date_key'] = df_w['date'].apply(lambda x: int(x.strftime('%Y%m%d')))

In [80]:
df.head(5)

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end_station_id end_station_name  end_station_latitude  ...  bike_id  \
0            3209     Brunswick St             40.724176  ...    24647   
1            3213   Van Vorst Park             40.718489  ...    24605   
2            3213   Van Vorst Park             40.718489  ...    24689   
3            3203    Hamilton Park             40.727596  ...    24693   
4            3210   Pershing Field             40.742677  ...    24573   

    user_type birth_year  gender  id   age  trip_duration_minutes  \
0  Subscriber     1964.0       2   0  52.0                   6.03   
1  Subscriber     1962.0       1   1  54.0                   3.33   
2  Subscriber     1962.0       2   2  54.0                   3.37   
3  Subscriber     1984.0       1   3  32.0                   4.13   
4    Customer        NaN       0   4   NaN                  15.05   

   trip_duration_hours  trip_flag  date_key  
0                 0.10       True  20160101  
1                 0.06       True  20160101  
2                 0.06       True  20160101  
3                 0.07       True  20160101  
4                 0.25       True  20160101  

[5 rows x 21 columns]

In [81]:
df_w.head()

name       date   awnd  prcp  snow  \
0  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01  12.75   0.0   0.0   
1  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   9.40   0.0   0.0   
2  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03  10.29   0.0   0.0   
3  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04  17.22   0.0   0.0   
4  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   9.84   0.0   0.0   

   snwd  tavg  tmax  tmin  date_key  
0   0.0    41    43    34  20160101  
1   0.0    36    42    30  20160102  
2   0.0    37    47    28  20160103  
3   0.0    32    35    14  20160104  
4   0.0    19    31    10  20160105

## Checking the information

We need to optimize the gestion of the table information

In [84]:
df.head()

trip_duration          start_time           stop_time  start_station_id  \
0            362 2016-01-01 00:02:52 2016-01-01 00:08:54              3186   
1            200 2016-01-01 00:18:22 2016-01-01 00:21:42              3186   
2            202 2016-01-01 00:18:25 2016-01-01 00:21:47              3186   
3            248 2016-01-01 00:23:13 2016-01-01 00:27:21              3209   
4            903 2016-01-01 01:03:20 2016-01-01 01:18:24              3195   

  start_station_name  start_station_latitude  start_station_longitude  \
0      Grove St PATH               40.719586               -74.043117   
1      Grove St PATH               40.719586               -74.043117   
2      Grove St PATH               40.719586               -74.043117   
3       Brunswick St               40.724176               -74.050656   
4            Sip Ave               40.730743               -74.063784   

   end_station_id end_station_name  end_station_latitude  ...  bike_id  \
0            3209     Brunswick St             40.724176  ...    24647   
1            3213   Van Vorst Park             40.718489  ...    24605   
2            3213   Van Vorst Park             40.718489  ...    24689   
3            3203    Hamilton Park             40.727596  ...    24693   
4            3210   Pershing Field             40.742677  ...    24573   

    user_type birth_year  gender  id   age  trip_duration_minutes  \
0  Subscriber     1964.0       2   0  52.0                   6.03   
1  Subscriber     1962.0       1   1  54.0                   3.33   
2  Subscriber     1962.0       2   2  54.0                   3.37   
3  Subscriber     1984.0       1   3  32.0                   4.13   
4    Customer        NaN       0   4   NaN                  15.05   

   trip_duration_hours  trip_flag  date_key  
0                 0.10       True  20160101  
1                 0.06       True  20160101  
2                 0.06       True  20160101  
3                 0.07       True  20160101  
4                 0.25       True  20160101  

[5 rows x 21 columns]

- We need if exist duplicated columns 

In [86]:
df.duplicated(subset = 'bike_id').value_counts()

True     247006
False       566
Name: count, dtype: int64

We can se that we have 247006 duplicated rows, it's strange. It could exist multiple bikers how rent a bike multiple days.
We only will send a report about that

We will reset the index of all the tables

In [89]:
df['id'].reset_index()
df_w['id'] = df_w.index

# Create tables 

### date_t table

In [92]:
df_date.head()

date  id  date_key  year  month  day
0 2016-01-01   0  20160101  2016      1    1
1 2016-01-02   1  20160102  2016      1    2
2 2016-01-03   2  20160103  2016      1    3
3 2016-01-04   3  20160104  2016      1    4
4 2016-01-05   4  20160105  2016      1    5

We well put all the information in order 

In [94]:
date_t = pd.DataFrame({
    'id': None,
    'date_key': df_date['date_key'],
    'date': df_date['date'],
    'year': df_date['year'],
    'month': df_date['month'],
    'day': df_date['day']  # O lo que corresponda aquí
})
date_t.reset_index(drop = True, inplace = True)
date_t['id'] = date_t.index

In [95]:
date_t.head()

id  date_key       date  year  month  day
0   0  20160101 2016-01-01  2016      1    1
1   1  20160102 2016-01-02  2016      1    2
2   2  20160103 2016-01-03  2016      1    3
3   3  20160104 2016-01-04  2016      1    4
4   4  20160105 2016-01-05  2016      1    5

### start_station table

In [97]:
start_station = pd.DataFrame({
    'id' : None,
    'start_id' : df['start_station_id'].copy(True),
    'start_station_name' : df['start_station_name'].copy(True),
    'start_station_latitude' : df['start_station_latitude'].copy(True),
    'start_station_latitude' : df['start_station_latitude'].copy(True),
    'start_station_longitude' : df['start_station_longitude'].copy(True)
})
start_station.reset_index(drop = True, inplace = True)
start_station['id'] = start_station.index

In [98]:
start_station.head()

id  start_id start_station_name  start_station_latitude  \
0   0      3186      Grove St PATH               40.719586   
1   1      3186      Grove St PATH               40.719586   
2   2      3186      Grove St PATH               40.719586   
3   3      3209       Brunswick St               40.724176   
4   4      3195            Sip Ave               40.730743   

   start_station_longitude  
0               -74.043117  
1               -74.043117  
2               -74.043117  
3               -74.050656  
4               -74.063784

We have duplicated information, we will drop some rows

In [100]:
start_station.drop_duplicates(subset = ['start_id'], inplace = True)
start_station.reset_index(drop = True, inplace = True)
start_station['id'] = start_station.index

In [101]:
start_station.head()

id  start_id start_station_name  start_station_latitude  \
0   0      3186      Grove St PATH               40.719586   
1   1      3209       Brunswick St               40.724176   
2   2      3195            Sip Ave               40.730743   
3   3      3211         Newark Ave               40.721525   
4   4      3187          Warren St               40.721124   

   start_station_longitude  
0               -74.043117  
1               -74.050656  
2               -74.063784  
3               -74.046305  
4               -74.038051

### end_station table

In [103]:
end_station = pd.DataFrame({
    'id' : None,
    'end_id' : df['end_station_id'].copy(True),
    'end_station_name' : df['end_station_name'].copy(True),
    'end_station_latitude' : df['end_station_latitude'].copy(True),
    'end_station_longitude' : df['end_station_longitude'].copy(True),
})
end_station.reset_index(drop = True, inplace = True)
end_station['id'] = end_station.index

In [104]:
end_station.head()

id  end_id end_station_name  end_station_latitude  end_station_longitude
0   0    3209     Brunswick St             40.724176             -74.050656
1   1    3213   Van Vorst Park             40.718489             -74.047727
2   2    3213   Van Vorst Park             40.718489             -74.047727
3   3    3203    Hamilton Park             40.727596             -74.044247
4   4    3210   Pershing Field             40.742677             -74.051789

We need to drop de duplicated values to

In [106]:
end_station.drop_duplicates(subset = ['end_id', 'end_station_name', 'end_station_latitude', 'end_station_longitude'], inplace = True)
end_station.reset_index(drop = True, inplace = True)
end_station['id'] = end_station.index

In [107]:
end_station.head()

id  end_id  end_station_name  end_station_latitude  end_station_longitude
0   0    3209      Brunswick St             40.724176             -74.050656
1   1    3213    Van Vorst Park             40.718489             -74.047727
2   2    3203     Hamilton Park             40.727596             -74.044247
3   3    3210    Pershing Field             40.742677             -74.051789
4   4    3214  Essex Light Rail             40.712774             -74.036486

### rider_info table

In [109]:
rider_info = pd.DataFrame({
    'id' : None,
    'biker_id' : df['bike_id'].copy(True),
    'user_type' : df['user_type'].copy(True),
    'birth_year' : df['birth_year'].copy(True),
    'gender': df['gender'].copy(True),
    'age' : df['age'].copy(True)
}) 
rider_info.reset_index(drop = True, inplace = True)
rider_info['id'] = rider_info.index

In [110]:
rider_info.head()

id  biker_id   user_type  birth_year  gender   age
0   0     24647  Subscriber      1964.0       2  52.0
1   1     24605  Subscriber      1962.0       1  54.0
2   2     24689  Subscriber      1962.0       2  54.0
3   3     24693  Subscriber      1984.0       1  32.0
4   4     24573    Customer         NaN       0   NaN

### rider table

In [112]:
rider = pd.DataFrame({
    'id' : None,
    'biker_id' : df['bike_id'].reset_index(drop=True),
    'date_key' : df['date_key'].reset_index(drop=True),
    'start_station_id' : start_station['start_id'].reset_index(drop=True),
    'end_station_id' : end_station['end_id'].reset_index(drop=True),
    'trip_duration_seconds' : df['trip_duration'].reset_index(drop=True),
    'trip_duration_minutes' : df['trip_duration_minutes'].reset_index(drop=True),
    'trip_duration_hours' : df['trip_duration_hours'].reset_index(drop=True),
    'rider_info_id' : rider_info['id'].reset_index(drop=True),
    'trip_flag' : df['trip_flag'].reset_index(drop=True)
})
rider.reset_index(drop = True, inplace = True)
rider['id'] = rider.index

In [113]:
rider.head()

id  biker_id  date_key  start_station_id  end_station_id  \
0   0     24647  20160101            3186.0          3209.0   
1   1     24605  20160101            3209.0          3213.0   
2   2     24689  20160101            3195.0          3203.0   
3   3     24693  20160101            3211.0          3210.0   
4   4     24573  20160101            3187.0          3214.0   

   trip_duration_seconds  trip_duration_minutes  trip_duration_hours  \
0                    362                   6.03                 0.10   
1                    200                   3.33                 0.06   
2                    202                   3.37                 0.06   
3                    248                   4.13                 0.07   
4                    903                  15.05                 0.25   

   rider_info_id  trip_flag  
0              0       True  
1              1       True  
2              2       True  
3              3       True  
4              4       True

### Weather table

In [115]:
df_w.head()

name       date   awnd  prcp  snow  \
0  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-01  12.75   0.0   0.0   
1  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-02   9.40   0.0   0.0   
2  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-03  10.29   0.0   0.0   
3  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-04  17.22   0.0   0.0   
4  NEWARK LIBERTY INTERNATIONAL AIRPORT, NJ US 2016-01-05   9.84   0.0   0.0   

   snwd  tavg  tmax  tmin  date_key  id  
0   0.0    41    43    34  20160101   0  
1   0.0    36    42    30  20160102   1  
2   0.0    37    47    28  20160103   2  
3   0.0    32    35    14  20160104   3  
4   0.0    19    31    10  20160105   4

In [116]:
weather = pd.DataFrame({
    'id' : df_w['id'].reset_index(drop=True, inplace = True),
    'date_key' : df_w['date_key'].reset_index(drop=True),
    'date' : df_w['date'].reset_index(drop=True),
    'awnd' : df_w['awnd'].reset_index(drop=True),
    'prcp' : df_w['prcp'].reset_index(drop=True),
    'snow' : df_w['snow'].reset_index(drop=True),
    'tavg' : df_w['tavg'].reset_index(drop=True),
    'tmax' : df_w['tmax'].reset_index(drop=True),
    'tmin' : df_w['tmin'].reset_index(drop=True)
})

In [117]:
weather.head()

id  date_key       date   awnd  prcp  snow  tavg  tmax  tmin
0  None  20160101 2016-01-01  12.75   0.0   0.0    41    43    34
1  None  20160102 2016-01-02   9.40   0.0   0.0    36    42    30
2  None  20160103 2016-01-03  10.29   0.0   0.0    37    47    28
3  None  20160104 2016-01-04  17.22   0.0   0.0    32    35    14
4  None  20160105 2016-01-05   9.84   0.0   0.0    19    31    10

# PostgresSQL Connection

In [119]:
#Conection with own Database
#Data
username = 'username'
password = 'password'
host = 'host'
port = 'port'
database = 'database'

In [120]:
#create the engine of SQLAlchemy
engine = sa.create_engine(f'postgresql://{username}:{password}@{host}:{port}/{database}')

In [195]:
con = engine.connect()
#date_t.to_sql('date_t', con, if_exists= 'append', index=False, chunksize=10000)
#start_station.to_sql('start_station', con, if_exists= 'append',index=False, chunksize=10000)
#end_station.to_sql('end_station', con, if_exists= 'append',index=False, chunksize=10000)
#rider_info.to_sql('rider_info', con, if_exists= 'append',index=False, chunksize=10000)
rider.to_sql('rider', con, if_exists= 'append',index=False, chunksize=10000)
weather.to_sql('end_table', con, if_exists= 'append',index=False, chunksize=10000)
con.close()